pdr_backend/sim/sim_engine.py

```diff
@@ -242,65 +264,99 @@ def run_one_iter(self, test_i: int, mergedohlcv_df: pl.DataFrame):
             self.sim_plotter.save_state(self.st, d, is_final_state)
 
     @enforce_types
+    # pylint: disable=too-many-return-statements
     def sim_trader(
         self,
         curprice: float,
         pred_up,
         pred_down,
         conf_up: float,
         conf_down: float,
+        high: float,
+        low: float,
     ) -> float:
         """
         @description
             Simulate trader's actions based on predictions and confidence levels.
             If trader has an open position, it will close it if the prediction
             changes. If trader has no open position, it will open a position if
-            the prediction is strong enough.
+            the prediction is strong enough. Also, close the position if the price
+            hits the stop loss or take profit levels.
 
         @arguments
             curprice -- current price of the token
             pred_up -- prediction that the price will go up
             pred_down -- prediction that the price will go down
             conf_up -- confidence in the prediction that the price will go up
             conf_down -- confidence in the prediction that the price will go down
+            high -- highest price reached during the period
+            low -- lowest price reached during the period
+
 
         @return
             profit -- profit made by the trader in this iteration
         """
 
+        def close_long_position(sell_price: float) -> float:
+            tokcoin_amt_send = self.position_size
+            usd_received = self._sell(sell_price, tokcoin_amt_send)
+            self.position_open = ""
+            profit = usd_received - self.position_worth
+            return profit
+
+        def close_short_position(buy_price: float) -> float:
+            usdcoin_amt_send = self.position_size * buy_price
+            self._buy(buy_price, usdcoin_amt_send)
+            self.position_open = ""
+            profit = self.position_worth - usdcoin_amt_send
+            return profit
+
```

Comment:
Do these need to be methods nested within this outer method? It adds complexity to the outer method. Instead, what if they were defined below, as first-class methods of SimEngine?